identity-server/src/lib.rs

```diff
@@ -58,3 +113,165 @@ impl RouterConfig {
 async fn root() -> &'static str {
 	"uwu hewwo this api is under constwuction"
 }
+
+/// Spawns a tokio task for handling HTTP requests and redirecting them to their https
+/// counterpart.
+pub async fn spawn_http_redirect_server(
+	http_port: u16,
+	https_port: NonZeroU16,
+) -> Result<(
+	tokio::task::JoinHandle<color_eyre::Result<()>>,
+	tokio::sync::oneshot::Sender<()>,
+	NonZeroU16,
+)> {
+	let listener = bind_listener(http_port).await?;
+	let local_addr = listener.local_addr().unwrap();
+	info!("HTTP redirect server listening on {local_addr}",);
+
+	let redirect = move |Host(host): Host, uri: Uri| async move {
+		match remap_http_to_https_url(&host, uri, https_port) {
+			Ok(uri) => Ok(Redirect::permanent(&uri.to_string())),
+			Err(error) => {
+				tracing::warn!(%error, "failed to convert URI to HTTPS");
+				Err(axum::http::StatusCode::BAD_REQUEST)
+			}
+		}
+	};
+	let serve_fut = axum::serve(listener, redirect.into_make_service())
+		.into_future()
+		.map(|r| r.wrap_err("HTTP redirect server crashed"));
+
+	let (tx, rx) = tokio::sync::oneshot::channel();
+	let task_handle = tokio::spawn(async move {
+		tokio::select! {
+			result = serve_fut => result,
+			_ = rx => {
+				info!("killing HTTP redirect server due to shutdown signal");
+				return Ok(());
+			}
+		}
+	});
+	Ok((task_handle, tx, local_addr.port().try_into().unwrap()))
+}
```

Comment:
@lyuma pointed out that this is actually asking for a security issue. Tomorrow I will remove this logic and simplify everything.